connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connect.execution
+
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
+
+import org.apache.spark.connect.proto.ExecutePlanResponse
+import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler.{SparkListener, SparkListenerJobEnd, SparkListenerJobStart, SparkListenerStageCompleted, SparkListenerTaskEnd, SparkListenerTaskStart}
+
+/**
+ * A listener that tracks the execution of jobs and stages for a given set of tags. This is used
+ * to track the progress of a job that is being executed through the connect API.
+ *
+ * The listener is instantiated once for the SparkConnectService and then used to track all the
+ * current query executions.
+ */
+private[connect] class ConnectProgressExecutionListener extends SparkListener with Logging {
+
+  /**
+   * A tracker for a given tag. This is used to track the progress of an operation is being
+   * executed through the connect API.
+   */
+  class ExecutionTracker(val tag: String) {
+
+    class StageInfo(
+        val stageId: Int,
+        var numTasks: Int,
+        var completedTasks: Int = 0,
+        var inputBytesRead: Long = 0,
+        var completed: Boolean = false) {
+
+      val lock = new Object
+      def update(i: StageInfo => Unit): Unit = {
+        lock.synchronized {
+          i(this)
+        }
+      }
+
+      def toProto(): ExecutePlanResponse.ExecutionProgress.StageInfo = {
+        ExecutePlanResponse.ExecutionProgress.StageInfo
+          .newBuilder()
+          .setStageId(stageId)
+          .setNumTasks(numTasks)
+          .setNumCompletedTasks(completedTasks)
+          .setInputBytesRead(inputBytesRead)
+          .setDone(completed)
+          .build()
+      }
+    }
+
+    // The set of jobs that are being tracked by this tracker. We always only add to this list
+    // but never remove. This is to avoid concurrency issues.
+    private[ConnectProgressExecutionListener] var jobs: Set[Int] = Set()
```

Comment:
nit: `private[this]` is the narrowest scope we can use here.

Reply:
Fixing it.

Reply:
Actually, this does not work here because `this` is not the right scope and gives compile errors.